node/primitives/src/disputes/mod.rs

```diff
@@ -221,26 +221,16 @@ impl SignedDisputeStatement {
 		};
 
 		let data = dispute_statement.payload_data(candidate_hash, session_index);
-		let signature = Keystore::sign_with(
-			&**keystore,
-			ValidatorId::ID,
-			&validator_public.clone().into(),
-			&data,
-		)?;
-
-		let signature = match signature {
-			Some(sig) =>
-				sig.try_into().map_err(|_| KeystoreError::KeyNotSupported(ValidatorId::ID))?,
-			None => return Ok(None),
-		};
-
-		Ok(Some(Self {
-			dispute_statement,
-			candidate_hash,
-			validator_public,
-			validator_signature: signature,
-			session_index,
-		}))
+		let signature = keystore
+			.sr25519_sign(ValidatorId::ID, validator_public.as_inner_ref(), &data)?
```

Comment:
Everywhere we assume a key is of a certain type it would be nice to have a static assertion here that the key is actually sr25519, e.g. something like this should work (you could wrap it in a macro if you want):

```rust
        const _: () = {
            let id = ValidatorId::CRYPTO_ID.0;
            if !(id[0] == b's' && id[1] == b'r' && id[2] == b'2' && id[3] == b'5') {
                panic!();
            }
        };
```

Reply:
The type is statically checked because of validator_public parameter. 
This doesn't compile if for example I change the call to `ecdsa_sign` or `ed25519_sign`

Reply:
Okay, I wasn't sure that's the case. Fair enough!